src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DerSequenceReader.cs

```diff
@@ -0,0 +1,262 @@
+using System;
+using System.Diagnostics;
+using System.Numerics;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Internal.Cryptography.Pal
+{
+    /// <summary>
+    /// Reads data encoded via the Distinguished Encoding Rules for Abstract
+    /// Syntax Notation 1 (ASN.1) data.
+    /// </summary>
+    internal class DerSequenceReader
+    {
+        private readonly byte[] _data;
+        private readonly int _end;
+        private int _position;
+
+        internal int ContentLength { get; private set; }
+
+        private DerSequenceReader(bool startAtPayload, byte[] data)
+        {
+            Debug.Assert(startAtPayload, "This overload is only for bypassing the sequence tag");
+            Debug.Assert(data != null, "Data is null");
+
+            _data = data;
+            _position = 0;
+            _end = data.Length;
+
+            ContentLength = data.Length;
+        }
+
+        internal DerSequenceReader(byte[] data)
+            : this(data, 0, data.Length)
+        {
+        }
+
+        internal DerSequenceReader(byte[] data, int offset, int length)
+        {
+            _data = data;
+            _end = offset + length;
+
+            Debug.Assert(data != null, "Data is null");
+            Debug.Assert(offset >= 0, "Offset is negative");
+            Debug.Assert(length > 2, "Length is too short");
+            Debug.Assert(data.Length >= offset + length, "Array is too short");
+
+            _position = offset;
+            EatTag(DerTag.Sequence);
+            ContentLength = EatLength();
+        }
+
+        internal static DerSequenceReader CreateForPayload(byte[] payload)
+        {
+            return new DerSequenceReader(true, payload);
+        }
+
+        internal bool HasData
+        {
+            get { return _position < _end; }
+        }
+
+        internal byte PeekTag()
+        {
+            return _data[_position];
+        }
+
+        internal void SkipValue()
+        {
+            EatTag((DerTag)PeekTag());
+            int contentLength = EatLength();
+            _position += contentLength;
+        }
+
+        internal int ReadInteger()
+        {
+            byte[] integerBytes = ReadIntegerBytes();
+
+            // integerBytes is currently Big-Endian, need to reverse it for
+            // Little-Endian to pass into BigInteger.
+            Array.Reverse(integerBytes);
+            BigInteger bigInt = new BigInteger(integerBytes);
+            return (int)bigInt;
```

Comment:
Are we expecting that integerBytes will just be 4-bytes long, or are we expecting this to wrap around?  If the former, could we just use BitConverter instead of going through a BigInteger?